src/dbnode/storage/index.go

```diff
@@ -2414,3 +2388,56 @@ func (shards dbShards) IDs() []uint32 {
 	}
 	return ids
 }
+
+// blocksIterStackAlloc is a stack allocated block iterator, ensuring no
+// allocations per query.
+type blocksIterStackAlloc struct {
+	blocks      []blockAndBlockStart
+	queryRanges xtime.Ranges
+	idx         int
+}
+
+func newBlocksIterStackAlloc(
+	blocks []blockAndBlockStart,
+	queryRanges xtime.Ranges,
+) blocksIterStackAlloc {
+	return blocksIterStackAlloc{
+		blocks:      blocks,
+		queryRanges: queryRanges,
+		idx:         -1,
+	}
+}
+
+func (i blocksIterStackAlloc) Next() (blocksIterStackAlloc, bool) {
```

Comment:
This took a few readthroughts to properly understand, could you add a comment describing what we're doing here? E.g. "Next() moves the iterator to the next block it owns that covers part of the given query range. This differs from other iterators, as it returns an incremented iterator and a boolean indicating if blocks exist. This returned iterator should be used in future operations, as Next() does not mutate or increment the original iterator"